### PR TITLE
feat(model): add option to return raw query

### PR DESCRIPTION
**Pull request by seromenho**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [ ] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [ ] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [ ] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### Description of change
Usage
```javascript
User.beforeFind(() => {
  console.log('I will be skipped');
});

const sql = await User.findAll({getRawSql: true})
console.log(sql); //  {query: "SELECT ....."}
```

Implements https://github.com/sequelize/sequelize/issues/2325

I'll come back to this and improve the PR description and add some tests if someone confirms this is a good direction.
Thanks

### CODE REVIEW

**Comment by jdgriffith** [comment]
This would be a great addition that our organization would use immediately. Any chance we can get feedback on whether this is a good approach?

**Review by dleo9307 — APPROVED** [review]

**Comment by tjwhitfield-lilly** [comment]
@dleo9307 What is the timeframe on this getting merged / released?

File: src/dialects/abstract/query-interface.js [hunk 1]
```diff
@@ -975,8 +975,13 @@ class QueryInterface {
 
   async select(model, tableName, optionsArg) {
     const options = { ...optionsArg, type: QueryTypes.SELECT, model };
+    const query = this.QueryGenerator.selectQuery(tableName, options, model);
```
Comment (baweinbe) [inline src/dialects/abstract/query-interface.js:978]: looks like this should be `const query = this.queryGenerator.selectQuery(tableName, options, model);`

**Review by viraldudhat — APPROVED** [review]

**Comment by jdgriffith** [comment]
Do I dare hope this is about to make it into sequelize?! 😍

**Comment by sdepold** [comment]
if it is WIP: Can we maybe convert it into a draft PR?

**Comment by sdepold** [comment]
fwiw: I just rebased. main is now including some bits and pieces about typescript. it shouldn't make a huge diff

File: lib/hooks.js [hunk 2]
```diff
@@ -90,7 +90,10 @@ const Hooks = {
 
   async runHooks(hooks, ...hookArgs) {
     if (!hooks) throw new Error('runHooks requires at least 1 argument');
-
+    // Skip hooks if we are only getting raw SQL
+    if (hookArgs[0] && typeof hookArgs[0] === 'object' && hookArgs[0].getRawSql) {
```
Comment (Keimeno) [inline lib/hooks.js:94]: Hooks can also modify the resulting query, such as `beforeFind`. I think it's best if we extend the `hookTypes` object, and add a new option to each hook that we want to run, when getting the raw SQL query

File: lib/dialects/abstract/query-interface.js [hunk 3]
```diff
@@ -949,11 +949,12 @@ class QueryInterface {
 
   async select(model, tableName, optionsArg) {
     const options = { ...optionsArg, type: QueryTypes.SELECT, model };
+    const query = this.queryGenerator.selectQuery(tableName, options, model);
+    if (options.getRawSql) {
```
Comment (Keimeno) [inline lib/dialects/abstract/query-interface.js:953]: I'm unsure if `getRawSql` is the best name for this option, I would probably go with something like `queryOnly`, `getQuery`, `query` or something else.

**Review by Keimeno — CHANGES_REQUESTED** [review]
Thanks for your contribution @seromenho, and sorry for the late response!

I've marked two things that I would change about this PR.

There's also a problem inside our `lib/model.js`, where we make use of the `select` method. https://github.com/sequelize/sequelize/blob/407137822a62897f7366980acd7eeceb443601b9/lib/model.js#L1754-L1770

We have a problem here, because our result is processed further, we will have to check for these scenarios as well.

Also, I'm unsure if we should limit this option for `findAndCountAll`, `findAll` and `findOne` queries, but it might as well make sense to separate this into multiple PRs

**Comment by ephys** [comment]
One note I have about this PR is that we should avoid having different return types for our functions. It makes typing them and using them with TypeScript much more complicated.

Instead, let's add a new method that does this. Maybe `Model.getFindQuery`, `Model.findAll.getQuery()`, or let's keep it as a method on the Query Builder? #394

**Comment by piotr-pawlowski** [comment]
When do you plan to merge this PR? We really need that feature.

**Comment by cr0mbly** [comment]
I know there was talks about making the `QueryGenerator` public, but it's been a few years. If this isn't going to be the case we can institute the above PR. Would love to use this natively.

**Comment by jdgriffith** [comment]
A part of me is sad to look back and see my first comment on this thread was four years ago. 

I think it's worth merging this PR, assuming it's ready or can be modified easily, as a stop gap until new features or v7 arrives.

**Comment by Nick-Riggs** [comment]
I've needed this for years for different projects, and I've always monkey-patched it. It may not cover every case, but something to the effect of:

```
    const realQuery = sequelize.query.bind(sequelize);

    sequelize.query = async function (sql: string, options: QueryOptions) {
        if (options?.returnSqlOnly) {
            return sql;
        }

        return realQuery(sql, options);
    }
```

I hope to see it in the framework proper someday as well.

**Comment by jdgriffith** [comment]
@Nick-Riggs Our team has had to do something similar because we have a dynamic query layer based on the sequelize ORM.

**Comment by cr0mbly** [comment]
@ephys this seems like a pretty big want/need from the community, are we able to look at this again?

I Just note again most ORMs I've dealt with have offered easy exposure of the SQL they're building within the ORM builder itself, it'd be nice if sequlize kept to that convention:

```
### Rails
Model.all.to_sql
>>> "SELECT `models`.* FROM `models`" 

### Django
Model.objects.all().query()
>>> "SELECT `models`.* FROM `models`"
```

**Pull request closed without merging**